### PR TITLE
Fix allocated Encoding memory initialization

### LINKED ISSUE #28438: [Arm64] System.Text.Tests.EncodingTest.TestRoundtrippingSpecificCodepageEncoding fails
```
xUnit.net console test runner (64-bit .NET Core)
Copyright (C) 2014 Outercurve Foundation.

Discovering: System.Text.Encoding.CodePages.Tests
Discovered:  System.Text.Encoding.CodePages.Tests
Starting:    System.Text.Encoding.CodePages.Tests
   System.Text.Tests.EncodingTest.TestRoundtrippingSpecificCodepageEncoding(encodingName: \"GB18030\", bytes: [129, 48, 137, 55, 129, ...], expected: \"Þßàáâã\") [FAIL]
      Assert.Equal() Failure
      Expected: Byte[] [129, 48, 137, 55, 129, ...]
      Actual:   Byte[] [129, 48, 137, 55, 129, ...]
      Stack Trace:
         /home/vmjenkins/workspace/Dotnet/build_corefx/src/System.Text.Encoding.CodePages/tests/EncodingCodePages.cs(535,0): at System.Text.Tests.EncodingTest.TestRoundtrippingSpecificCodepageEncoding(String encodingName, Byte[] bytes, String expected)
Finished:    System.Text.Encoding.CodePages.Tests

=== TEST EXECUTION SUMMARY ===
   System.Text.Encoding.CodePages.Tests  Total: 740, Errors: 0, Failed: 1, Skipped: 0, Time: 5.775s
```

### PR DESCRIPTION
**Pull request by tarekgh**
Fixes #28438
We allocate native memory using GlobalHAlloc which usually doesn’t zero out the allocated memory and that causes a problem when using this memory. We fixed a similar issue before in SBCS but looks we missed the DBCS one.

### CODE REVIEW

**Comment by tarekgh** [comment]
@krwq could you please have a look?

CC @sdmaclea

**Comment by tarekgh** [comment]
Link the related commit to this issue:
https://github.com/krwq/corefx/commit/02c66c3ef81745f383aa2251a101366628561584

**Comment by stephentoub** [comment]
A few questions:
- Why haven't we seen failures on other platforms?
- Why do we allocate native memory rather than a managed allocation?
- Do we need to do this zero'ing, or is there something we can change elsewhere such that we won't need it zero'd?

**Comment by tarekgh** [comment]
@stephentoub 

> Why haven't we seen failures on other platforms?

I believe we were just lucky.

> Why do we allocate native memory rather than a managed allocation?

This is a code which originally ported from the desktop. I recall there were 2 reasons for using native memory, the first is to allocate one time and use it across app domains. this reason is not applicable to the net core. the second reason is the rest of the code depend on the pointer value, so if we allocated managed pointer we'll have to change all referencing code to not assuming this is happening. otherwise, we'll have to pin this buffer and put more pressure on GC.

> Do we need to do this zero'ing, or is there something we can change elsewhere such that we won't need it zero'd?

Yes, we need to ensure the memory is zero-initialized. 

@sdmaclea 

I can switch to use Unsafe.InitBlockUnaligned if think this will be faster. I may try to measure that and if it is faster, I can switch to use it instead.

**Comment by sdmaclea** [comment]
> I can switch to use Unsafe.InitBlockUnaligned if think this will be faster

I'm not sure if it will be faster.  I saw span<> was using it.  And there were **stale** benchmarks on stackoverflow which seemed to indicate it would be faster.

File: src/System.Text.Encoding.CodePages/src/System/Text/BaseCodePageEncoding.cs [hunk 1]
```diff
@@ -340,5 +340,27 @@ internal unsafe void CheckMemorySection()
                 LoadManagedCodePage();
             }
         }
+
+        // Method assumes that memory pointer is aligned
+        internal static unsafe void ZeroMemAligned(byte* buffer, int count)
```
Comment (danmosemsft) [inline src/System.Text.Encoding.CodePages/src/System/Text/BaseCodePageEncoding.cs]: Do we not have any API already that we can use to zero memory? This does not seem special in any way.

@GrabYourPitchforks ?

**Comment by tarekgh** [comment]
Looks Unsafe.InitBlockUnaligned is really faster. on my AMD64 machine, I am seeing it is faster by 28.75% than our implementation. I'll switch to use Unsafe.InitBlockUnaligned
```
   System.Globalization.Tests.Perf_ZeroMemory.ManualZeroMemAligned     | Duration | msec |     2      | 7689.468 |  10.817 | 7681.819 | 7697.117
   System.Globalization.Tests.Perf_ZeroMemory.UnsafeInitBlockUnaligned | Duration | msec |     2      | 5478.959 |  13.631 | 5469.320 | 5488.597
```

File: src/System.Text.Encoding.CodePages/src/System/Text/BaseCodePageEncoding.cs (hunk 1 above)
Comment (tarekgh) [reply src/System.Text.Encoding.CodePages/src/System/Text/BaseCodePageEncoding.cs]: we have Unsafe.InitBlockUnaligned and I am going to switch using it.

**Comment by tarekgh** [comment]
CC @weshaggard as added a dependency on System.Runtime.CompilerServices.Unsafe.

File: src/System.Text.Encoding.CodePages/src/System/Text/SBCSCodePageEncoding.cs [hunk 2]
```diff
@@ -92,7 +71,7 @@ protected override unsafe void LoadManagedCodePage()
                 const int CodePageNumberSize = 4;
                 int bytesToAllocate = UnicodeToBytesMappingSize + BytesToUnicodeMappingSize + CodePageNumberSize + iExtraBytes;
                 byte* pNativeMemory = GetNativeMemory(bytesToAllocate);
-                ZeroMemAligned(pNativeMemory, bytesToAllocate);
+                Unsafe.InitBlockUnaligned(pNativeMemory, 0, (uint) bytesToAllocate);
```
Comment (krwq) [inline src/System.Text.Encoding.CodePages/src/System/Text/SBCSCodePageEncoding.cs]: nit: space after `(uint)`

**Comment by tarekgh** [comment]
@weshaggard could you please look at last commit? especially the file packageindex.json? the changes there generated by compiling the target 

```
/t:UpdatePackageIndex
```

I built successfully all packages but want to ensure the changes looks good to you before I merge if we get clean CI.

**Comment by weshaggard** [comment]
>@weshaggard could you please look at last commit? especially the file packageindex.json? the changes there generated by compiling the target

Looks like the tool just reformatted the file (cc @erictstj). If you want to ensure you aren't doing anything else I would suggest undoing the file and only applying the changes for the Unsafe library.

**Comment by tarekgh** [comment]
@weshaggard 

> If you want to ensure you aren't doing anything else I would suggest undoing the file and only applying the changes for the Unsafe library.

I don't prefer this because next time anyone running the target /t:UpdatePackageIndex will run into the same case. if you feel strongly about it, I'll do what you have suggested.

@ericstj do you have any concern here?

**Comment by ericstj** [comment]
> @ericstj do you have any concern here?

No, but I had thought that @joperezr was applying a stable sort to these so that we don't see churn.  We just need to tack on an OrderBy to this guy: https://github.com/dotnet/buildtools/blob/ba9e9d01729a9131c30f9c074450eae0409f21c1/src/Microsoft.DotNet.Build.Tasks.Packaging/src/PackageIndex.cs#L725

**Comment by tarekgh** [comment]
ok, will go ahead and merge this and will log an issue to track the ordering.

**Pull request merged**